engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/state/RunStateNode.java

```diff
@@ -1,18 +1,163 @@
+  Stateful doSingletonNewKeyCached(
+      SingletonMap state,
+      Object _this,
+      Object key,
+      Object local_state,
+      Thunk computation,
+      @Cached("key") Object cachedNewKey,
+      @Cached("state.getKey()") Object cachedOldKey,
+      @Cached(value = "buildSmallKeys(cachedNewKey, cachedOldKey)", dimensions = 1)
+          Object[] newKeys) {
+    SmallMap localStateMap = new SmallMap(newKeys, new Object[] {local_state, state.getValue()});
+    Stateful res = thunkExecutorNode.executeThunk(computation, localStateMap, false);
+    Object newStateVal = ((SmallMap) res.getState()).getValues()[1];
+    return new Stateful(new SingletonMap(cachedOldKey, newStateVal), res.getValue());
+  }
+
+  @Specialization
+  Stateful doSingletonNewKeyUncached(
+      SingletonMap state, Object _this, Object key, Object local_state, Thunk computation) {
+    return doSingletonNewKeyCached(
+        state,
+        _this,
+        key,
+        local_state,
+        computation,
+        key,
+        state.getKey(),
+        buildSmallKeys(key, state.getKey()));
+  }
+
+  Object[] buildSmallKeys(Object k1, Object k2) {
+    return new Object[] {k1, k2};
+  }
+
+  @Specialization(
+      guards = {"key == cachedNewKey", "state.getKeys() == cachedOldKeys", "index == NOT_FOUND"})
+  Stateful doMultiNewKeyCached(
+      SmallMap state,
+      Object _this,
+      Object key,
+      Object local_state,
+      Thunk computation,
+      @Cached("key") Object cachedNewKey,
+      @Cached(value = "state.getKeys()", dimensions = 1) Object[] cachedOldKeys,
+      @Cached("state.indexOf(key)") int index,
+      @Cached(value = "buildNewKeys(cachedNewKey, cachedOldKeys)", dimensions = 1)
+          Object[] newKeys) {
+    Object[] newValues = new Object[newKeys.length];
+    System.arraycopy(state.getValues(), 0, newValues, 1, cachedOldKeys.length);
+    newValues[0] = local_state;
+    SmallMap localStateMap = new SmallMap(newKeys, newValues);
+    Stateful res = thunkExecutorNode.executeThunk(computation, localStateMap, false);
+    SmallMap resultStateMap = (SmallMap) res.getState();
+    Object[] resultValues = new Object[cachedOldKeys.length];
+    System.arraycopy(resultStateMap.getValues(), 1, resultValues, 0, cachedOldKeys.length);
+    return new Stateful(new SmallMap(cachedOldKeys, resultValues), res.getValue());
+  }
+
+  @Specialization(
+      guards = {"key == cachedNewKey", "state.getKeys() == cachedOldKeys", "index != NOT_FOUND"})
+  Stateful doMultiExistingKeyCached(
+      SmallMap state,
+      Object _this,
+      Object key,
+      Object local_state,
+      Thunk computation,
+      @Cached("key") Object cachedNewKey,
+      @Cached(value = "state.getKeys()", dimensions = 1) Object[] cachedOldKeys,
+      @Cached("state.indexOf(key)") int index) {
+    Object[] newValues = new Object[cachedOldKeys.length];
+    System.arraycopy(state.getValues(), 0, newValues, 0, cachedOldKeys.length);
+    newValues[index] = local_state;
```

Comment:
As above, if it doesn't hurt performance, this and the second part (`resultValues[index] = ...` and the rest) maybe could be changed into some function `Object[] updateArray(Object[] array, int index, Object newValue)`. I think the function would be much easier to read and understand then.